source/extensions/filters/udp/dns_filter/dns_filter_resolver.cc

```diff
@@ -0,0 +1,76 @@
+#include "extensions/filters/udp/dns_filter/dns_filter_resolver.h"
+
+#include "common/network/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace UdpFilters {
+namespace DnsFilter {
+
+void DnsFilterResolver::resolveExternalQuery(DnsQueryContextPtr context,
+                                             const DnsQueryRecord* domain_query) {
+  if (active_query_ != nullptr) {
```

Comment:
In the case of a retry, it's just making sure that the active timer gets cancelled.   The `active_query_` is a pointer to a timer.  This is returned by the call to `resolve()`.

Again, having multiple queries in one buffer is supported by the protocol, but it is not widely (if at all) implemented.   I haven't found any cases where tools support this.

Reply:
I am not referring to multiple queries in a single request.

I am referring to multiple dns requests from different clients, while some are pending external resolution. Isn't there an issue there since you will overwrite the state of the external resolver when it happens?

Reply:
It does look that way.  I can add the context to the lambda so when the resolution completes it's still available for the filter to respond to the client.

Reply:
I am either reading the code incorrectly or not able to communicate effectively ☺️

Let me try asking differently:

- There is 1 resolver object per filter.
- The resolver has 1 member variable called external context.

Given the above, if there is a pending external resolution, this member variable has a valid pointer, which is read and invoked during the callback after the resolution completes.

Think of what could happen if the external query is still outstanding and a new request comes in for another external query.

Wouldn't the member variables be reset? 

I think the problem with this repro script is the external resolution is not slow enough.
Can you inject a sleep of 2 seconds in the external resolution code path and retry your repro script?

Reply:
I'm not sure that injecting a longer sleep during external resolutions will elicit the condition you are referring to.  Any delay would be absorbed by `dig` so it would have to be introduced outside of this script.

Let me change this.  I see what you mean, though I am not able to repro it.

Reply:
I meant to inject the sleep in the envoy filter code during external resolution. Not the script.

Reply:
I would let @mattklein123  comment on this aspect - We can definitely setup a call to discuss more, but just want to eliminate if my understanding of the envoy threading model is incorrect with Matt taking a quick look at this part

Reply:
@sumukhs is correct, I don't think this code will work as is. Here is a summary of the UDP filter threading model:
1) There is 1 filter per listener per worker.
2) The filter can receive any number of UDP packets.
3) The filter will need to deal with concurrent UDP queries that have outstanding upstream DNS requests, probably by maintaining some type of set of outstanding requests, hashed by origin IP/port tuple. (It might be worth it to look at how the state management works in the UDP proxy filter.)

If it would be easier to setup a short call to discuss we can do that next week.

Reply:
I'll take a  look at the UDP proxy and rework this class.